File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        // https://leetcode.com/problems/longest-common-subsequence/discuss/348884/C%2B%2B-with-picture-O(nm)
        int n1 = text1.size(), n2 = text2.size();
        vector<vector<int>> m(n1 + 1, vector<int>(n2 + 1));
        // Do this to avoid [i-1][j-1] checks
        for (int i = 1; i <= n1; i++) {
            for (int j = 1; j <= n2; j++) {
                  if (text1[i - 1] == text2[j - 1]) {
                      m[i][j] = m[i - 1][j - 1] + 1;
                      // Add on from prevs
                  } else {
                      m[i][j] = max(m[i - 1][j], m[i][j - 1]);
                      // Maintain max
                  }
            }
        }
        return m[n1][n2];
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution(object):
    def __init__(self):
        self.d = {}
    
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """
        return self.memoize(text1,text2,len(text1)-1, len(text2)-1)
        
            
    def memoize(self, text1, text2, i1, i2):
        if (i1 < 0 or i2 < 0):
            return 0
        elif ((i1, i2) in self.d):
            return self.d[(i1, i2)]
        else:
            res = 0
            if (text1[i1] == text2[i2]):
                res = 1 + self.memoize(text1,text2,i1-1,i2-1)
            else:
                sub1 = self.memoize(text1,text2,i1-1,i2)
                sub2 = self.memoize(text1,text2,i1,i2-1)
                res = max(sub1, sub2)
                
            self.d[(i1, i2)] = res
            return res